### PR TITLE
shelf_router (without the redirect rules)

### PR DESCRIPTION
**Pull request by isoos**
I've left the redirect rules to a second round, and there may be some work to handle the pub client API or the cron jobs better.

### CODE REVIEW

File: app/lib/frontend/handlers/routes.dart [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+  @Route.get('/web/packages')
+  Future<Response> webPackages(Request request) =>
+      webPackagesHandlerHtml(request);
+
+  // ****
+  // **** Packages
+  // ****
+
+  @Route.get('/packages/<package>/versions/<version>')
```
Comment (jonasfj) [inline app/lib/frontend/handlers/routes.dart]: Note for later: we could probably also do a sub-router for `/packages/`.

File: app/lib/frontend/handlers/routes.dart [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+  Future<Response> robotsTxt(Request request) => robotsTxtHandler(request);
+
+  /// Renders the /sitemap.txt page
+  @Route.get('/sitemap.txt')
+  Future<Response> sitemapTxt(Request request) => siteMapTxtHandler(request);
+
+  /// Renders static assets
+  @Route.get('/static/<path|[^]*>')
+  Future<Response> staticAsset(Request request, String path) {
+    // TODO: pass-in the [path] parameter
```
Comment (jonasfj) [inline app/lib/frontend/handlers/routes.dart]: Careful to also handle the query string :)

File: app/lib/frontend/handlers/routes.dart [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+  /// Removing an existing uploader.
+  ///
+  /// DELETE /api/packages/<package-name>/uploaders/<uploader-email>
+  /// https://github.com/dart-lang/pub_server/blob/master/lib/shelf_pubserver.dart#L118-L123
+  @Route.delete('/api/packages/<package>/uploaders/<email>')
+  Future<Response> removeUploader(
+          Request request, String package, String email) async =>
+      _pubServerHandler(request);
+
```
[the remark this answers is not in the document]
Comment (isoos) [reply app/lib/frontend/handlers/routes.dart]: This indicates it was with an `/api/` prefix:
https://github.com/dart-lang/pub_server/blob/master/lib/shelf_pubserver.dart#L67-L75
Nevertheless, added the other route to the same method.

I'm planning to implement an integration test with the fake_pub_server and your script before this lands, so we merge it only if that test passes.

[the remark this answers is not in the document]
Comment (jonasfj) [reply app/lib/frontend/handlers/routes.dart]: Oh, yes, but that's not what is implemented in the client.. The json payloads also contains URLs and these are never read either.. look at the request logs and grep for `.tar.gz` :(

[the remark this answers is not in the document]
Comment (jonasfj) [reply app/lib/frontend/handlers/routes.dart]: We probably need to migrate the client to use URLs from the json..

**Comment by isoos** [comment]
rebased to include integration tests.

**Pull request merged**